org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JDTStandaloneFileUtils.java

```diff
@@ -0,0 +1,194 @@
+/*******************************************************************************
+ * Copyright (c) 2018 Microsoft Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Microsoft Corporation - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.jdt.ls.core.internal;
+
+import java.net.URI;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.codehaus.plexus.util.StringUtils;
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IFolder;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.Path;
+import org.eclipse.jdt.core.ICompilationUnit;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.ls.core.internal.preferences.PreferenceManager;
+
+public class JDTStandaloneFileUtils {
+	public static final String PATH_SEPARATOR = "/";
+	public static final String PERIOD = ".";
+	public static final String SRC = "src";
+	public static final String WORKSPACE_LINK = "_";
+
+	public static ICompilationUnit getFakeCompilationUnit(URI uri, IProgressMonitor monitor) {
+		if (uri == null || !"file".equals(uri.getScheme()) || !uri.getPath().endsWith(".java")) {
+			return null;
+		}
+		java.nio.file.Path path = Paths.get(uri);
+		//Only support existing standalone java files
+		if (!java.nio.file.Files.isReadable(path)) {
+			return null;
+		}
+
+		IPath workspaceRoot = getWorkspaceRoot(ResourceUtils.filePathFromURI(uri.toString()));
+		if (workspaceRoot != null) {
+			return getFakeCompilationUnitForInsideStandaloneFile(uri, workspaceRoot, monitor);
+		} else {
+			return getFakeCompilationUnitForExternalStandaloneFile(uri, monitor);
+		}
+	}
+
+	public static IPath getWorkspaceRoot(IPath filePath) {
+		PreferenceManager manager = JavaLanguageServerPlugin.getPreferencesManager();
+		Collection<IPath> rootPaths = manager.getPreferences().getRootPaths();
+		for (IPath rootPath : rootPaths) {
+			if (rootPath.isPrefixOf(filePath)) {
+				return rootPath;
+			}
+		}
+
+		return null;
+	}
+
+	public static String getWorkspaceInvisibleProjectName(IPath workspacePath) {
+		String fileName = workspacePath.toFile().getName();
+		String projectName = fileName + "_" + Integer.toHexString(workspacePath.toPortableString().hashCode());
```

Comment:
"_" => WORKSPACE_LINK

Reply:
Here `_` is just a simple separator. But the WORKSPACE_LINK symbol represents a special folder. Don't like to replace it here.